### DISCUSSION TITLE
Date and Time

### DISCUSSION THREAD

**Discussion by Sineos** [phoddie/node-red-mcu#62] [Q&A] [answered]
I have successfully build my first highly complex flow:

![image](https://user-images.githubusercontent.com/4459312/209632883-1c96e6b4-42c0-4b53-8094-b5762c800cb3.png)

Hardly surprising the return value is the Unix epoch since no date and time is initialized on the MCU. 
What is the correct way to do so? Is there a "NTP node" that could pull the current time?

**Comment by phoddie** [comment, marked as answer]
@Sineos – congratulations! Very complex, indeed. ;)

The date/time of the MCU is in initialized in a couple of different ways:

- On a debug build, the computer sends the current time to the MCU when the debugging connection (xsbug) is established. This is automatic and behind the scenes. It is very convenient for development. If you run a debug build disconnected from a computer, it cannot get the time that way, of course. Then you are in the same situation as the next point.
- On release and instrumentation builds, there is no debugging connection so the time needs to come from somewhere else.
   - On boards with a battery-backed Real-time Clock component the host initializes the time as part of set-up before Node-MCU starts. We see RTCs in our commercial client work, but they are fairly rare in development boards.
   - If your project connects to Wi-Fi at start-up, it can optionally also retrieve the current time from an NTP server. That is common enough, so the option is built into our ESP32 host: it just needs to be configured. When building from the command line, define your preferred NTP server:
   ```mcconfig -d -m sntp=pool.ntp.org```
   I believe this is also supported in @ralphwetzel's Node-RED MCU Plug-in.

Maybe this is common enough that we should default Node-RED MCU Edition projects to always configure an NTP server? That would just mean adding this to manifest_runtime.json:

```json
	"config": {
		"sntp": "pool.ntp.org"
	}
```

Of course, a project could still override the default NTP server in that case.

**Reply by ralphwetzel** [reply]
> * I believe this is also supported in @ralphwetzel's Node-RED MCU Plug-in.

It is:
<img width="382" alt="image" src="https://user-images.githubusercontent.com/16342003/209713204-36a4afb3-8482-49b8-a98b-5649f72f0b71.png">

**Reply by phoddie** [reply]
Of course! As an aside to this topic, what do you think about making the label of that field "Arguments / config" (or something like that)? The idea is to make it more clear that the field applies to the `"config"` section of the manifest (command line arguments to `mcconfig` are just one way to define the config).

**Reply by ralphwetzel** [reply]
The [mcconfig docu](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/tools/tools.md#arguments) calls them `config arguments` ... which I think is a good term. Thank you for this proposal!

<img width="389" alt="image" src="https://user-images.githubusercontent.com/16342003/209727519-c8ffd78e-614b-48dd-8a8f-c6797a059111.png">

**Reply by phoddie** [reply]
Perfect. Great to match the docs. Thank you!

**Reply by Sineos** [reply]
Awesome, works like a charm.
This is really a fantastic project. Looking forward to seeing it evolve

**Reply by phoddie** [reply]
FYI – the default configuration now defines an NTP server, so it isn't necessary to define one unless you want to use a specific server. To keep project start-up as fast as possible, it does not contact the NTP server if the  time is already set (for example, by the debugger connection).
